tatt.1
.TH TATT 1
.SH NAME
tatt (is an) arch testing tool.
.SH SYNOPSIS
.B tatt
\fI<options> [suboptions]\fB

.SH DESCRIPTION
Tatt is an arch-testing tool written for automation of frequently repeated tasks.

.SH OPTIONS
.TP 
\fB-h, --help\fI
Show this help message and exit.
.TP
\fB-d, --depend\fI
Determine stable rdeps.
.TP
\fB-u, --use, --usecombis\fI
Determine use flag combinations.
.TP
\fB-f INFILE, --file=INFILE\fI
Input File containing packages.
.TP
\fB-j JOBNAME, --jobname=JOBNAME\fI
Name for the job, prefix of output files.
.TP
\fB-b BUGNUM, --bug=BUGNUM\fI
Do the full program for a given stable request bug
.TP
\fB-s SUCCBUGNUM, --success=SUCCBUGNUM\fI
Comment that the program was successfully tested
.TP
\fB-r RESOLVENUM, --resolve=RESOLVENUM\fI
Resolve the given bugnumber, needs a message
.TP
\fB-c, --close\fI
Resolve the given bugnumber with closing it, needs to be combined with -r
.TP
\fB-m RESOLVEMESSAGE, --message=RESOLVEMESSAGE\fI
Message for bug resolution.
.TP
\fB-k, --keywording\fI
Search for keywording packages, needs to be combined with -f
.TP
\fB-v, --verbose\fI
Print informative output.

.SH USAGE
.TP
tatt -b300000 -j myjob
Work on a stable bug no 300000. This will unmask the package and
create five shell scripts.  One is for automated testing of USE-flag
combinations, one is for testing of reverse dependencies, one is for
committing the new keywords to CVS, one is for leaving a message on
the bug, and finally one is for cleaning up.

-j specifies a jobname which will be a prefix for the scripts that
tatt produces, if it is left empty the bugnumber will be used

.TP
tatt -f myPackageFile -b bugnumber
This will open the file myPackageFile, look for all atoms in it, and
write scripts that test/commit all packages.  If -j is omitted the
filename is used.  The bugnumber is necessary for the commit script.

.TP
tatt -r bugnum -m "x86 stable, Thanks xyz" 
Removes your arch from the CC field of the bug and adds the message (assuming everything was committed and we want to resolve the bug).
.br
-c switch additionally closes the bug.
.SH FILES
~/.tatt
.br
/usr/share/tatt/templates/

.SH SEE ALSO
gatt(8)
.br
https://github.com/gentoo/tatt

.SH COPYRIGHT 
This software is licensed under the GNU General Public License, version 2 (GPLv2). A copy of this license is available at http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.

.SH BUGS 
Forks and pull requests are welcome on GitHub. https://github.com/gentoo/tatt

.SH AUTHOR
tatt is written and maintained by Thomas Kahle <tomka@gentoo.org>


tatt.py
#!/usr/bin/python

from gentooPackage import gentooPackage as gP
from tool import unique

from subprocess import *
import sys
import re
import random
import os

# from configobj
from configobj import ConfigObj
from validate  import Validator


########### Generate a global config obj #################

config = ConfigObj("dot-tatt-example", configspec="dot-tatt-spec")

# This validator will also do type conversion according to the spec file!
validator = Validator()
result = config.validate(validator)

if result != True:
    print 'Config file validation failed!'
    sys.exit(1)

## Generate stable rdeps ###
def stableredeps (atom):
    """
    Find packages with stable versions which depend on atom
    We query the tinderbox at http://tinderbox.dev.gentoo.org/misc/dindex/
    for this purpose.
    The result is a list of pairs of package atoms and a list of necessary useflags
    """

    #Todo, this will not work with atoms that specify verisons
    import urllib
    tinderbox = 'http://tinderbox.dev.gentoo.org/misc/dindex/'
    download = urllib.urlopen(tinderbox + atom).read()
    if not re.search("404 - Not Found", download) == None:
        return []
    packlist = download.rstrip().split("\n")
    # Split at : to see if useflags are necessary
    splitlist2 = [p.split(":") for p in packlist]
    # Fill with empty useflags if nothing is given:
    splitlist = []
    for s in splitlist2:
        if len(s) == 1:
            splitlist.append([s[0],[" "]])
        else:
            splitlist.append([s[0],s[1].split("+")])
    d = dict([])
    for s in splitlist:
        # Saves necessary useflags under package names, removing duplicates.
        d[gP(s[0]).packageName()] = s[1]
    outlist2 = [[k, d[k]] for k in d.keys()]
    outlist = []
    for o in outlist2:
        # We are calling eix for each package to work around issues with --stable:
        eixcall = ["eix", "--stable", "--only-names", "--exact", o[0]]
        p2 = Popen(eixcall, stdout=PIPE)
        out = p2.communicate()[0]
        if out == '': continue
        else : outlist.append(o)
    return outlist
    
#############################

## Useflag Combis ##
def findUseFlagCombis (atom):
    """
    Generate combinations of use flags to test
    """
    
    uses=Popen('equery -C uses '+atom+' | cut -f 1 | cut -c 2-40 | xargs',
               shell=True, stdout=PIPE).communicate()[0]
    uselist=uses.split()
    # The uselist could have duplicates due to slot-conditional
    # output of equery
    uselist=unique(uselist)
    for i in config["ignoreprefix"]:
        uselist=[u for u in uselist if not re.match(i,u)]

    if len(uselist) > 4:
        # More than 4 use flags, generate 16 random strings and everything -, everything +
        s = 2**(len (uselist))
        random.seed()
        swlist = [random.randint(0, s-1) for i in range (16)]
        swlist.append(0)
        swlist.append(s-1)
        swlist.sort()
        swlist = unique(swlist)
    else:
        # 4 or less use flags. Generate all combinations
        swlist = range(2**len(uselist))

    usecombis=[]
    for sw in swlist:
        mod = []
        for pos in range(len(uselist)):
            if ((2**pos) & sw):
                mod.append("")
            else:
                mod.append("-")
        usecombis.append(zip(mod, uselist))

    usecombis = [["".join(uf) for uf in combi] for combi in usecombis]

    # Merge everything to as USE="" string
    return ["USE=\""+" ".join(uc)+ "\"" for uc in usecombis]
#####################################################


#### Write useflagcombiscript ########
def writeusecombiscript(atom):
    # Show or build with diffent useflag combis
    usecombis = findUseFlagCombis (atom)
    outfilename = (atom.split("/")[1] + "-useflagtest.sh")
    reportname = (atom.split("/")[1] + ".report")
    if os.path.isfile(outfilename):
        print ("WARNING: Will overwrite " + outfilename)
    outfile = open(outfilename, 'w')
    outfile.write("#!/bin/sh" + '\n')
    for uc in usecombis:
        outfile.write("if " + uc + " emerge -1v " + atom + "; then " + '\n')
        outfile.write("  echo \"" + uc.replace("\"","\'") + " succeeded \" >> " + reportname + "; " + '\n')
        outfile.write("else echo \"" + uc.replace("\"", "\'") + " failed \" >> " + reportname + '; \nfi; \n')
    if options.feature_test:
        # Test once with system enabled useflags in the end:
        outfile.write ("FEATURES=\"test\" emerge -1v " + atom)
    outfile.close()
    print ("Build commands written to " + outfilename)
    return 0
######################################


### Write rdepcombiscript ############
def writerdepscript(atom):
    # We are checking for stable rdeps:
    rdeps = stableredeps (atom)
    if len(rdeps) == 0:
        print "No stable rdeps"
        return 0
    if len(rdeps) > 20:
        print "More than 20 stable rdeps, sampling 20"
        rdeps = random.sample(rdeps, 20)
    outfilename = (atom.split("/")[1] + "-rdeptest.sh")
    reportname = (atom.split("/")[1] + ".report")
    if os.path.isfile(outfilename):
        print ("WARNING: Will overwrite " + outfilename)
    outfile = open(outfilename,'w')
    outfile.write("#!/bin/sh" + '\n')
    for r in rdeps:
        call = ""
        if options.feature_test:
            call = "FEATURES=\"test\" "
        call = (call + "USE=\"" + " ".join([s for s in r[1] if not s[0] == "!"]) + " ")
        call = (call + " ".join(["-" + s[1:] for s in r[1] if s[0] == "!"]))
        call = (call + "\" emerge -1v " + r[0])
        st = ("if " + call + "; then \n")
        st = (st + "echo \""+call.replace("\"","\'") + "\" succeeded >> " + reportname + ";\n")
        st = (st + "else echo \""+call.replace("\"","\'") + "\" failed >> " + reportname + ";\nfi;\n")
        outfile.write(st)
    outfile.close()
    print ("Rdep build commands written to " + outfilename)
    return 0
######################################

#######Write report script############
def writereportscript (bugnum, atom):
    outfilename = (atom.split("/")[1] + "-success.sh")
    reportname = (atom.split("/")[1] + ".report")
    if os.path.isfile(outfilename):
        print ("WARNING: Will overwrite " + outfilename)
    outfile = open(outfilename,'w')
    outfile.write("#!/bin/sh" + '\n')
    outfile.write("if grep failed " + reportname + " >> /dev/null; then echo Failure found;\n")
    outfile.write("else bugz modify " + bugnum + ' -c' + "\"" +config['successmessage'] + "\";\n")
    outfile.write("fi;")
    outfile.close()
    print ("Success Report script written to " + outfilename)
    return 0

######### Main program starts here ###############

### USAGE and OPTIONS ###
from optparse import OptionParser

parser=OptionParser()
parser.add_option("-d", "--depend",
                  help="Determine stable rdeps",
                  dest="depend",
                  action="store_true",
                  default = False)
parser.add_option("-u", "--use" "--usecombis",
                  help="Determine use flag combinations",
                  dest="usecombi",
                  action="store_true",
                  default = False)
parser.add_option("-f", "--file", "-o",
                  help="Outfile name",
                  dest="fileprefix",
                  action="store",
                  default="tatt-test.sh"
                  )
parser.add_option("-p", "--pretend", 
                  help="Print things to stdout instead of doing them",
                  action="store_true",
                  default=False
                  )
parser.add_option("-t", "--test",
                  help="run emerge commands with FEATURES=\"test\"",
                  dest="feature_test",
                  action="store_true",
                  default = True)
parser.add_option("-b", "--bug",
                  help="do the full program for a given stable request bug",
                  dest="bugnum",
                  action="store")
parser.add_option("-s", "--success",
		  help="Comment that the program was successfully tested",
                  dest="succbugnum",
		  action="store")

(options,args) = parser.parse_args()

if (Popen(['whoami'], stdout=PIPE).communicate()[0].rstrip() == 'root'):
    isroot=True
else:
    print "You're not root!"
    isroot=False

## -s and a bugnumber was given ?
if options.succbugnum:
    print "Reporting success for bug number " + options.succbugnum
    retcode = call(['bugz', 'modify', options.succbugnum, '-c', config['successmessage']])
    if retcode == 0:
        print "Success!";
        exit (0)
    else:
        print "Failure commenting on Bugzilla"
        exit(1)

## -b and a bugnumber was given ?
if options.bugnum:
    print "Working on bug number " + options.bugnum
    bugraw = Popen(['bugz', 'get', options.bugnum, '-n', '--skip-auth'], stdout=PIPE).communicate()[0]
    if not re.search('[Ss][Tt][Aa][Bb]', bugraw):
        print "Does not look like a stable request bug !"
        print bugraw
        # Let's not exit here, maybe we still want to work on the bug
#        exit (1)
    bugdata = bugraw.split("\n")
    
    # Todo: Check if this is really an atom
    # atomre = re.compile("=?[^\s:,;<>]+/\S+-[0-9]?\S+[0-9][^\s:,;<>]?[a-z]*")
    atomre = re.compile(config['atom-regexp'])
    for l in bugdata:
        m = atomre.search(l)
        if m == None: continue
        atom = m.group(0)
        break
    # Remove a leading =
    p = gP(atom)
    print "Found the following package atom : " + p.packageString()
    # Splitting the atom to get the package name:
    if isroot:
        # If we are root, then we can write to package.keywords
        try:
            keywordfile=open("/etc/portage/package.keywords/arch", 'r+')
        except IOError:
            # create an empty file, this should be beautified
            keywordfile=open("/etc/portage/package.keywords/arch", 'w')
            keywordfile.write(" ")
            keywordfile.close()
            keywordfile=open("/etc/portage/package.keywords/arch", 'r+')

        # Test if keywordfile already contains the atom
        if re.search(p.packageString(), keywordfile.read()):
            print "Package atom already in package.keywords"
        else:
            keywordfile.write("\n" + p.packageString() + "\n")
            print "Appended package to /etc/portage/package.keywords/arch"
        keywordfile.close()
    else:
        print "You are not root, your unmaskstring would be:"
        print ("\n" + p.packageString() + "\n")
    ## Write the scripts
    writeusecombiscript(p.packageName())
    writerdepscript(p.packageName())
    writereportscript(options.bugnum, p.packageName())
    exit (0)

## If we arrive here then a package atom should be given
try:
    atom = args[0]
except IndexError:
    print "Please call with package atom as argument"
    exit (1)

if options.depend:
    writerdepscript(atom)

if options.usecombi:
    writeusecombiscript(atom)

## That's all folks ##
